### DISCUSSION TITLE
Where to add pages for different tenants

### DISCUSSION THREAD

**Discussion by LRathje** [OrchardCMS/OrchardCore#8572] [Q&A]
Hi
I have setup a multi tenant decoupled install for using Razor mainly following this description: https://docs.orchardcore.net/en/dev/docs/guides/decoupled-cms/
I see the App_Data folder now contains the different sites succesfully. I do want to make my own MVC-like pages, but can't figure out which structure to setup for pages so that I don't get the pages for the Default site for all sites?
Is it a matter of:
- Making some kind of pages or view structure within the root-pages folder?
- Add a module per site and manage it like MVC from there?
- Place the pages or views within the App_Data folder? But here I can't include them in the project.
- Or something else?
Please if someone can push me in the right direction. What ever I do, I have only managed to render the root page folder index.cshtml file for all tenants..
Best regards
Lars

**Comment by LRathje** [comment]
Let me frame the question a bit more precise after having digged into it more:
- I setup a multi tenant environment of decoupled sites using OrchardCore.Application.Cms.Core.Targets
- One of these tenants is named eg Tenant2 with urlprefix ten2
- How can I do my own controllers, views and layouts for this Tentant2 independent of Tenant1, Tenant3 etc?
- Ideally I want to use a separate project for Tenant2, but utilize the dashboard functionality and database content of the single main OrchardCore install
Best regards
Lars

**Reply by deanmarcussen** [reply]
If you are seeing the two tenants as having very different functionality, i.e. controllers / pages / views, you may want consider making two different themes, one for each tenant (could also been done as modules, but probably theming is more suitable).

It is more common, in decoupled, to create different content type views, and an option could be to create a different content type per tenant. 

But I suspect you want more theming like functionality than that. You can still use the concept of theming with a decoupled approach, by just activating the theme you want on the tenant you want.

It would be reasonably common for this requirement to have a theme per tenant, and possibly a module per tenant, with the different bits of functionality that are not shared, and maybe a common module, for functionality that is shared.

**Reply by jtkech** [reply]
@LRathje 

Normally you need to create modules / features, then for a given tenant the controller is only triggered if the related feature is enabled, the view being shared by all tenants

A module has at least one feature, so you can have a module per controller, and in the module startup you can configure a custom route for the related controller.

But a module can contain multiple features, so you can have only one module for multiple controllers each being tied to a different feature, we have a feature attribute for this, and if you need custom routes you need separate startup decorated with he same feature attribute.

Finally you can do that from the app that can behave as a module, for this you need to define features as it is done in a module Manifest.cs file, and use our OrchardCoreBuilder helpers that allows to define e.g. a `Configure()` that will get called as it was defined in a module startup, this if you need custom routes.

Note: I think you will need custom routes as the default ones uses the assembly names, so in case of the app's module the app assembly name, or use the route attribute for your controllers, hmm but here the app level routing may match first ans then you would not run in the context of OrchardCore, meaning in a tenat context composed of modules, and so on.

**Reply by jtkech** [reply]
Oops, @deanmarcussen didn't see your answer ;)

**Reply by LRathje** [reply]
Thanks for quick replies both of you :)
I'm not that well into OchardCore yet, so I will spend some time to study and understand your recommendations and revert.
Anyway, what really started this, is that I see the standard themes provided as great for the admin, but very limited for the tenants. And I love using razor, so I thought maybe you could provide a project module for each tenant that gives me that flexibility to create my own razor and bootstrap views. Plus I have a few custom features that I want to present on the pages as widgets, modules or whatever.. Just to explain my reasons :).

**Reply by LRathje** [reply]
Thanks again - here is what I went on with:
The main app is a core web app using OrchardCore.Application.Cms.Core.Targets with a simple startup of services.AddOrchardCms and app.UseOrchardCore. I enable the tenants and can access admin for the default as well as for each tenant like this tenantdomain/admin.
For tenants where I need a MVC structure, I then add a core mvc app and set outputtype as library using OrchardCore.Module.Targets with an empty startup and no program. This module is then included in the project and inside the specific tenants admin I enable my new module. After this I can access my MVC through e.g tenantdomain/modulename/home/index. I can also use route attributes on my controllers an then just access like this tenantdomain/routeattribute.
My next step is now to let these controllers and views access the orchardcorecms contents and display it in the views.
This is a promising structure for me and the main trick is to activate modules from specific tenants admins.
Thanks both of you - I believe this is getting close to what you suggested @jtkech.
Best regards
Lars

**Reply by jtkech** [reply]
Cool, good luck !!!
